File: java/Graphs/DFS.java
package java.Graphs; // Depth First Search

import java.util.*;

public class DFS {
  public List<Integer> dfsOfGraph(int V, ArrayList<ArrayList<Integer>> adj) {
    List<Integer> result = new ArrayList<>();
    boolean[] visited = new boolean[V];
    dfs(0, adj, visited, result);
    return result;
  }

  void dfs(int node, ArrayList<ArrayList<Integer>> adj, boolean[] visited, List<Integer> result) {
    visited[node] = true;
    result.add(node);
    for (int neighbor : adj.get(node)) {
      if (!visited[neighbor]) {
        dfs(neighbor, adj, visited, result);
      }
    }
  }
}

File: C++/Graphs/DFS.cpp
#include <bits/stdc++.h>
using namespace std;

void dfsHelper(int node, vector<int> adj[], vector<bool> &visited, vector<int> &dfs)
{
  visited[node] = true;
  dfs.push_back(node);

  for (int neighbor : adj[node])
  {
    if (!visited[neighbor])
      dfsHelper(neighbor, adj, visited, dfs);
  }
}

vector<int> dfsOfGraph(int V, vector<int> adj[])
{
  vector<bool> visited(V, false);
  vector<int> dfs;

  dfsHelper(0, adj, visited, dfs);

  return dfs;
}